test/test_rtc.cpp

```diff
@@ -0,0 +1,314 @@
+	};
+
+	auto handler = [&](peer_id const&, rtc_stream_init) {};
+
+	sig = std::make_shared<rtc_signaling>(io_context, &tor1, handler);
+
+	std::cout << "Generating " << requested_offers_count << " offers" << std::endl;
+	sig->generate_offers(requested_offers_count, offers_handler);
+
+	run_test();
+
+	sig->close();
+}
+
+void test_connectivity()
+{
+	std::shared_ptr<rtc_signaling> sig1, sig2;
+	bool endpoint1_connected = false;
+	bool endpoint2_connected = false;
+
+	auto answer_callback = [&](peer_id const&, rtc_answer const& answer) {
+		std::cout << "Signaling 2: Generated an answer" << std::endl;
+
+		std::cout << "Signaling 1: Processing the answer" << std::endl;
+		sig1->process_answer(answer);
+	};
+
+	auto offers_handler = [&](error_code const& ec, std::vector<rtc_offer> offers) {
+		TEST_CHECK(!ec);
+
+		std::cout << "Signaling 1: Generated " << int(offers.size()) << " offer(s)" << std::endl;
+		TEST_EQUAL(int(offers.size()), 1);
+
+		rtc_offer offer = offers[0];
+		offer.answer_callback = answer_callback;
+
+		std::cout << "Signaling 2: Processing the offer" << std::endl;
+		sig2->process_offer(offer);
+	};
+
+	auto handler1 = [&](peer_id const&, rtc_stream_init init) {
+		TEST_CHECK(init.peer_connection);
+		TEST_CHECK(init.data_channel);
+
+		std::cout << "Signaling 1: Endpoint is connected" << std::endl;
+		endpoint1_connected = true;
+
+		if(endpoint2_connected)
+		{
+			std::cout << "Test succeeded" << std::endl;
+			success = true;
+		}
+	};
+
+	auto handler2 = [&](peer_id const&, rtc_stream_init init) {
+		TEST_CHECK(init.peer_connection);
+		TEST_CHECK(init.data_channel);
+
+		std::cout << "Signaling 2: Endpoint is connected" << std::endl;
+		endpoint2_connected = true;
+
+		if(endpoint1_connected)
+		{
+			std::cout << "Test succeeded" << std::endl;
+			success = true;
+		}
+	};
+
+	sig1 = std::make_shared<rtc_signaling>(io_context, &tor1, handler1);
+	sig2 = std::make_shared<rtc_signaling>(io_context, &tor2, handler2);
+
+	std::cout << "Signaling 1: Generating 1 offer" << std::endl;
+	sig1->generate_offers(1, offers_handler);
+
+	run_test();
+
+	sig1->close();
+	sig2->close();
+}
+
+void test_stream()
+{
+	std::shared_ptr<rtc_signaling> sig1, sig2;
+	std::shared_ptr<rtc_stream> stream1, stream2;
+
+	std::vector<char> message(16*1024);
+	std::iota(message.begin(), message.end(), char(0));
+	std::random_device rd;
+	std::shuffle(message.begin(), message.end(), std::mt19937(rd()));
+
+	std::vector<char> message_buffer(message.size());
+	boost::asio::mutable_buffer read_buffer(message_buffer.data(), message_buffer.size());
+
+	bool written = false;
+	bool received = false;
+
+	auto answer_callback = [&](peer_id const&, rtc_answer const& answer) {
+		std::cout << "Signaling 2: Generated an answer" << std::endl;
+
+		std::cout << "Signaling 1: Processing the answer" << std::endl;
+		sig1->process_answer(answer);
+	};
+
+	auto offers_handler = [&](error_code const& ec, std::vector<rtc_offer> offers) {
+		TEST_CHECK(!ec);
```

Comment:
and it asserts that there is never an error, which (afaict) statically must be true. `ec` seems redundant